Common/cpp/Registries/NewestShadowNodesRegistry.cpp

```diff
@@ -0,0 +1,64 @@
+#include "NewestShadowNodesRegistry.h"
+
+#include <react/renderer/core/ComponentDescriptor.h>
+#include <react/renderer/core/ShadowNodeFragment.h>
+
+using namespace facebook::react;
+
+namespace reanimated {
+
+void NewestShadowNodesRegistry::set(
+    ShadowNode::Shared shadowNode,
+    Tag parentTag) {
+  map_[shadowNode->getTag()] = std::make_pair(shadowNode, parentTag);
+}
+
+bool NewestShadowNodesRegistry::has(
+    const ShadowNode::Shared &shadowNode) const {
+  return map_.find(shadowNode->getTag()) != map_.cend();
+}
+
+ShadowNode::Shared NewestShadowNodesRegistry::get(Tag tag) const {
+  const auto it = map_.find(tag);
+  return it != map_.cend() ? it->second.first : nullptr;
+}
+
+void NewestShadowNodesRegistry::update(ShadowNode::Shared shadowNode) {
+  const auto it = map_.find(shadowNode->getTag());
+  react_native_assert(it != map_.cend());
+  it->second.first = shadowNode;
+}
+
+void NewestShadowNodesRegistry::remove(Tag tag) {
+  if (map_.find(tag) == map_.cend()) {
+    return;
+  }
+
+  auto shadowNode = map_[tag].first;
+
+  while (shadowNode != nullptr) {
+    bool hasAnyChildInMap = false;
+    for (const auto &child : shadowNode->getChildren()) {
+      if (has(child)) {
+        hasAnyChildInMap = true;
+        break;
+      }
+    }
+
+    if (hasAnyChildInMap) {
```

Comment:
Imagine the following ShadowTree:
```mermaid
flowchart TD
    A((A)) --> B((B)) --> C((C))
    B((B)) --> D((D)) --> E((E)) --> F((F))
    D((D)) --> G((G))
```
Assume that `C` and `F` are animated views. In `NewestShadowNodesRegistry` we need to keep not only animated ShadowNodes but also its parents all the way up to the root. Thus, for `C` we need to store `B` and `A`, and for `F` we need to store `E`, `D`, `B` and `A`.

Now, imagine we remove view `F`. Consequently, we also need to remove `E` and `D` (because `G` is not an animated component). We cannot remove `B` because it is located on the path from `C` to the root.

Since we perform the cleanup in a bottom-up fashion, we can remove ShadowNode from the registry only if none of its children is present in the registry.

Once we remove `F` from the registry we move on to its parent. `E` has no children left in the registry so we can remove it as well. Similarly for `D`, neither `E` or `G` is present in the registry, so we remove it. For `B` we can see that its child `C` is present in the registry, so we break from the loop, leaving `B` untouched.

I believe this could be implemented better with reference counting using `shared_ptr` but I need some more time to think about it.